src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java

```diff
@@ -26,23 +26,65 @@
+ * a function descriptor. The method type, consisting of a set of 'carrier' types constitutes
+ * the Java side of a call to a foreign function, while the function descriptor constitutes
+ * the native side of the call. Memory layout attributes are used in the function descriptor
+ * to attach ABI classification meta-data to memory layouts, which are required for linking.
+ * Clients of this API should use the prepared memory layout constants found in this interface
+ * to create their function descriptor, based on the built-in types provided by the C language.</p>
+ *
+ * <p>Implementations of this interface support the following primitive carrier types:
+ * {@code byte}, {@code short}, {@code char}, {@code int}, {@code long}, {@code float},
+ * and {@code double}, as well as {@link MemoryAddress} for passing pointers, and
+ * {@link MemorySegment} for passing structs and unions. Finally, the {@link VaList}
+ * carrier type can be used to match the native {@code va_list} type.</p>
+ *
+ * <p>The function descriptor used in linking contains a memory layout to match each carrier type.
+ * There are some restrictions on the carrier type and memory layout combinations that are allowed:</p>
+ * <ul>
+ *   <li>If a primitve type is used as a carrier type, the corresponding
+ *   memory layout must be a {@code ValueLayout}, and the bit size of the layout must match that of the carrier type
+ *   (see {@link Integer#SIZE} and similar fields in other primitive wrapper classes).</li>
+ *
+ *   <li>If the carrier type is {@code MemoryAddress} or {@code VaList}, then the corresponding memory layout must be a
+ *   {@code ValueLayout}, and its bit size must match the platform's address size (see {@link MemoryLayouts#ADDRESS}).
+ *   For this purpose, {@link CLinker#C_POINTER} can  be used as a memory layout for {@code MemoryAddress} carriers,
+ *   and {@link CLinker#C_VA_LIST} can be used for {@code VaList} carriers.</li>
+ *
+ *   <li>If the carrier type is {@code MemorySegment}, then the corresponding memory layout must be a
+ *   {@code GroupLayout}</li>
+ * </ul>
+ *
+ * <p>Variadic functions, declared in C either with a trailing ellipses ({@code ...}) at the end of the formal parameter
+ * list or with an empty formal parameter list, are not supported directly. It is not possible to create a method handle
+ * that takes a variable number of arguments, and neither is it possible to create an upcall stub wrapping a method
+ * handle that accepts a variable number of arguments. However, for down calls only, it is possible to link a native
+ * variadic function by using a 'specialized' method type and function descriptor: for each argument that is to be
+ * passed as a variadic argument, an explicit carrier type and memory layout must be present in the method type and
+ * function descriptor when linking the function. Furthermore, the memory layouts of variadic arguments must
```

Comment:
Furthermore, as memory layouts corresponding to variadic arguments in a function descriptor must contain additional classification information, it is required that the function {@link #asVarArg(MemoryLayout)} is called on each parameter layout that is used to compose a specialized variadic function descriptor.

Reply:
This is not quite correct, as it must only be called on variadic arguments, not fixed ones. I went with:

Furthermore, as memory layouts corresponding to variadic arguments in a function descriptor must contain additional classification information, it is required that {@link #asVarArg(MemoryLayout)} is used to create the memory layouts for each parameter corresponding to a variadic arguments in a specialized function descriptor.